zebra-state/src/service/finalized_state/zebra_db.rs

```diff
@@ -81,52 +91,44 @@ impl ZebraDb {
             db: DiskDb::new(config, network),
         };
 
-        db.check_max_on_disk_tip_height();
+        db.spawn_format_change(config, network, format_change);
 
-        // We have to get this height before we spawn the upgrade task, because threads can take
-        // a while to start, and new blocks can be committed as soon as we return from this method.
-        let initial_tip_height = db.finalized_tip_height();
+        db
+    }
 
+    /// Launch any required format changes or format checks, and store their thread handle.
+    pub fn spawn_format_change(
+        &mut self,
+        config: &Config,
+        network: Network,
+        format_change: DbFormatChange,
+    ) {
         // Always do format upgrades & checks in production code.
-        if cfg!(test) && debug_skip_format_upgrades {
-            return db;
+        if cfg!(test) && self.debug_skip_format_upgrades {
+            return;
         }
 
-        // Start any required format changes, and do format checks.
-        //
-        // TODO: should debug_stop_at_height wait for these upgrades, or not?
-        if let Some(format_change) = format_change {
-            // Launch the format change and install its handle in the database.
-            //
-            // `upgrade_db` is a special clone of the database, which can't be used to shut down
-            // the upgrade task. (Because the task hasn't been launched yet,
-            // `db.format_change_handle` is always None.)
-            //
-            // It can be a FinalizedState if needed, or the FinalizedState methods needed for
-            // upgrades can be moved to ZebraDb.
-            let upgrade_db = db.clone();
-
-            let format_change_handle = format_change.spawn_format_change(
-                config.clone(),
-                network,
-                initial_tip_height,
-                upgrade_db,
-            );
-
-            db.format_change_handle = Some(format_change_handle);
-        } else {
-            // If we're re-opening a previously upgraded or newly created database,
-            // the database format should be valid.
-            // (There's no format change here, so the format change checks won't run.)
-            //
-            // Do the quick checks first, then the slower checks.
-            upgrade::add_subtrees::quick_check(&db);
-
-            DbFormatChange::check_for_duplicate_trees(db.clone());
-            upgrade::add_subtrees::check(&db);
-        }
-
-        db
+        // We have to get this height before we spawn the upgrade task, because threads can take
+        // a while to start, and new blocks can be committed as soon as we return from this method.
```

Comment:
It think it depends on exactly how the upgrade code is written.

For deduplicating trees if a duplicate tree for the new block wasn't written, then it won't appear in the iterator, and nothing happens. If it's a unique block then it will appear in the iterator and it will get kept.

For the subtrees it might overwrite the subtree from a new block with the same subtree data from the upgrade. Which is completely ok.

A partial upgrade is very similar to adding new blocks using the latest state code. In both cases the database is a mix of versions. So it should usually work, but we don't test it very well, because it's an edge case.